src/CSE.cpp

```diff
@@ -112,102 +101,44 @@ class GVN : public IRMutator {
     Expr mutate(const Expr &e) override {
         // Early out if we've already seen this exact Expr.
         {
-            map<Expr, int, ExprCompare>::iterator iter = shallow_numbering.find(e);
+            auto iter = shallow_numbering.find(e.get());
             if (iter != shallow_numbering.end()) {
                 number = iter->second;
-                internal_assert(entries[number].expr.type() == e.type());
-                return entries[number].expr;
+                auto &entry = entries[number];
+                return entry.expr;
             }
         }
 
-        // If e is a var, check if it has been redirected to an existing numbering.
-        if (const Variable *var = e.as<Variable>()) {
-            if (let_substitutions.contains(var->name)) {
-                number = let_substitutions.get(var->name);
-                internal_assert(entries[number].expr.type() == e.type());
-                return entries[number].expr;
-            }
-        }
-
-        // If e already has an entry, return that.
-        CacheType::iterator iter = numbering.find(with_cache(e));
-        if (iter != numbering.end()) {
-            number = iter->second;
-            shallow_numbering[e] = number;
-            internal_assert(entries[number].expr.type() == e.type());
-            return entries[number].expr;
-        }
-
-        // Rebuild using things already in the numbering.
-        Expr old_e = e;
+        // We haven't seen this exact Expr before. Rebuild it using
+        // things already in the numbering.
+        number = -1;
         Expr new_e = IRMutator::mutate(e);
 
-        // See if it's there in another form after being rebuilt
-        // (e.g. because it was a let variable).
-        iter = numbering.find(with_cache(new_e));
-        if (iter != numbering.end()) {
+        // 'number' is now set to the numbering for the last child of
+        // this Expr (or -1 if there are no children). Next we see if
+        // that child has an identical parent to this one.
+
+        auto &use_map = number == -1 ? leaves : entries[number].uses;
+        auto p = use_map.emplace(with_cache(new_e), (int)entries.size());
+        auto iter = p.first;
+        bool novel = p.second;
+        if (novel) {
+            // This is a never-before-seen Expr
+            number = (int)entries.size();
+            entries.emplace_back(new_e);
+            iter->second = number;
```

Comment:
This is not safe: iter refers to an item contained in `entries`, but the call to emplace_back could have reallocated the items in entries, making iter invalid. (Wish I could say I saw this from inspection, but ASAN is the hero here.)

Reply:
iter refers to an item in the use_map, which is merely pointed to by .... I'm passing maps around by value aren't I. Will fix.